components/builder-api-client/src/builder.rs

```diff
@@ -1426,6 +1430,45 @@ impl BuilderAPIClient {
     }
 }
 
+/// Retry API operations within reasonable bounds.
+/// Provide an Option with a custom back-off algorithm from the `retry` crate
+/// or a default Fibonacci algorithm will be used. Unrecoverable HTTP errors
+/// will not be retried.
+#[macro_export]
+macro_rules! retry_builder_api {
+    ($iter_opt:expr, $api_future:expr) => {
+        async {
+            let mut iterator = if let Some(i) = $iter_opt {
+                i
+            } else {
+                retry::delay::Fibonacci::from(API_RETRY_WAIT).take(API_RETRIES)
+            };
+            retry::retry_future!(iterator, async {
+                match $api_future.await.into() {
+                    Ok(_) => retry::OperationResult::Ok(()),
+                    Err(e @ api_client::Error::APIError(StatusCode::NOT_IMPLEMENTED, _)) => {
+                        println!("Unsupported package platform or architecture. Skipping!");
+                        retry::OperationResult::Ok(())
+                    }
+                    Err(e @ api_client::Error::APIError(StatusCode::NOT_FOUND, _)) => {
+                        retry::OperationResult::Err(e)
+                    }
+                    Err(e @ api_client::Error::APIError(StatusCode::UNAUTHORIZED, _)) => {
+                        retry::OperationResult::Err(e)
+                    }
+                    Err(e @ api_client::Error::APIError(StatusCode::FORBIDDEN, _)) => {
+                        retry::OperationResult::Err(e)
+                    }
+                    Err(e @ api_client::Error::APIError(StatusCode::UNPROCESSABLE_ENTITY, _)) => {
+                        retry::OperationResult::Err(e)
+                    }
+                    Err(e) => retry::OperationResult::Retry(e),
```

Comment:
Are there going to be specific errors you'll return in the future (when this is out of draft status) for each of these cases? As currently written, everything is basically the same, allowing you to collapse most of these match arms.

Reply:
These match arms remain are purposeful as such so that they return a `retry::OperationResult::Err(e)` which terminates the retry wrapper.  Every other error type returns a `retry::OperationResult::Retry(e)` which allows the retry wrapper to iterate to the next loop cycle. 

This is because I don't want to needlessly retry these HTTP errors that are effectively unrecoverable (404, 401, 403, 422).

Reply:
I wonder if we should skip retry for any error in the 4XX Client Errros range? (Or alternatively only retry HTTP errors in the 5XX Server Errors range?)

Reply:
@christophermaier I've been battling this error when I try your suggestion above:
```
error[E0382]: use of moved value
--> components/common/src/command/package/install.rs:843:9
|
843 | /         retry_builder_api!(async {
844 | |             ui.status(Status::Downloading, format!("{} for {}", ident, target))?;
845 | |             self.api_client
846 | |                 .fetch_package((ident.as_ref(), target),
...   |
850 | |                 .await
851 | |         }).await
| |          ^
| |          |
| |__________value moved here
|            value used here after move
|
= note: move occurs because value has type `habitat_api_client::error::Error`, which does not implement the `Copy` trait
= note: this error originates in a macro (in Nightly builds, run with -Z macro-backtrace for more info)
```  
:thinking: The Error enum doesn't implement Copy or Clone and doesn't seem to be easily changed due to all the `String` types within.

Reply:
I am not sure why the error is happening. The error message loses some of its usefulness passing through the macro. It looks like just reconstructing an `ApiError` works though. Something like:

```rust

Reply:
Ah yes! the `if` guard on the match arm was just what I was looking for to avoid the move!
Thanks @davidMcneil - added in https://github.com/habitat-sh/habitat/pull/7874/commits/879cbd572b6ac202a85b3d4b2fa201fd6326abe5